### PR TITLE
Improve TaskResult Rest handling

### PR DESCRIPTION
**Pull request by fviale**
- fix issues related to raw results (byte array), add a isRaw information to know that the result is in byte array format and should not be deserialized using an ObjectInputStream
- use Base64 conversion for TaskResult's serialized objects in dozer mappings
- Enable variable retrieval from the rest TaskResult (TaskResultData). Variables are accessible either in serialized/base64 format or in text format (toString)

### CODE REVIEW

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 20 issues
* MAJOR 4 major
* MINOR 16 minor

#### Top 10 issues

1. MAJOR [ObjectByteConverter.java#L79](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L79): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ObjectByteConverter.java#L210](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L210): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ByteArrayToBase64StringCustomConverter.java#L53](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/ByteArrayToBase64StringCustomConverter.java#L53): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [SerializableToStringCustomConverter.java#L62](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/SerializableToStringCustomConverter.java#L62): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MINOR [ObjectByteConverter.java#L215](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L215): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectByteConverter.java#L230](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L230): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [TaskResult.java#L79](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L79): Define and throw a dedicated exception instead of using a generic one. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00112)
1. MINOR [TaskResult.java#L135](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L135): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectUtility.java#L28](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L28): Remove this unused import 'java.io.ByteArrayInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [ObjectUtility.java#L30](https://github.com/ow2-proactive/scheduling/blob/ad0e0a55284d2f43ad93196520c9b06bcff6dbb8/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L30): Remove this unused import 'java.io.ObjectInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 20 issues
* MAJOR 4 major
* MINOR 16 minor

#### Top 10 issues

1. MAJOR [ObjectByteConverter.java#L79](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L79): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ObjectByteConverter.java#L210](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L210): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ByteArrayToBase64StringCustomConverter.java#L53](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/ByteArrayToBase64StringCustomConverter.java#L53): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [SerializableToStringCustomConverter.java#L62](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/SerializableToStringCustomConverter.java#L62): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MINOR [ObjectByteConverter.java#L215](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L215): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectByteConverter.java#L230](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L230): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [TaskResult.java#L78](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L78): Define and throw a dedicated exception instead of using a generic one. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00112)
1. MINOR [TaskResult.java#L134](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L134): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectUtility.java#L28](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L28): Remove this unused import 'java.io.ByteArrayInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [ObjectUtility.java#L30](https://github.com/ow2-proactive/scheduling/blob/ed02802744cb493e36bfc8cb8b408716baf6561a/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L30): Remove this unused import 'java.io.ObjectInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 20 issues
* MAJOR 4 major
* MINOR 16 minor

#### Top 10 issues

1. MAJOR [ObjectByteConverter.java#L79](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L79): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ObjectByteConverter.java#L210](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L210): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [ByteArrayToBase64StringCustomConverter.java#L53](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/ByteArrayToBase64StringCustomConverter.java#L53): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MAJOR [SerializableToStringCustomConverter.java#L62](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/dto/SerializableToStringCustomConverter.java#L62): Return an empty array instead of null. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1168)
1. MINOR [ObjectByteConverter.java#L215](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L215): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectByteConverter.java#L230](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java#L230): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [TaskResult.java#L78](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L78): Define and throw a dedicated exception instead of using a generic one. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS00112)
1. MINOR [TaskResult.java#L134](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/task/TaskResult.java#L134): Refactor this method to throw at most one checked exception instead of: java.io.IOException, java.lang.ClassNotFoundException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [ObjectUtility.java#L28](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L28): Remove this unused import 'java.io.ByteArrayInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [ObjectUtility.java#L30](https://github.com/ow2-proactive/scheduling/blob/cae1f85b761bf34ce9194968b7d273e4f1e6ca00/rest/rest-api/src/main/java/org/ow2/proactive_grid_cloud_portal/utils/ObjectUtility.java#L30): Remove this unused import 'java.io.ObjectInputStream'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java [hunk 1]
```diff
@@ -70,6 +75,9 @@
     public static final byte[] objectToByteArray(Object obj, boolean compress) throws IOException {
         ByteArrayOutputStream baos = null;
         ObjectOutputStream oos = null;
+        if (obj == null) {
```
Comment (marcocast) [inline common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java]: Shouldn't we let the method throws a NullPointerException in this case? and also all the similar null checks in this pull request?

Comment (fviale) [inline common/common-api/src/main/java/org/ow2/proactive/utils/ObjectByteConverter.java]: We want to be able to convert a null result object altogether. In that case, it should not be transformed as a empty byte array, but keep the null quality. The result can be null, if for example an exception occurred during the task

**Pull request merged**